core/src/avm2/bytearray.rs

```diff
@@ -33,316 +34,210 @@ impl ByteArrayStorage {
-        let ushort_bytes = match self.endian {
-            Endian::Big => val.to_be_bytes(),
-            Endian::Little => val.to_le_bytes(),
-        };
-        self.write_bytes(&ushort_bytes);
+    pub fn read_boolean(&self) -> Result<bool, Error> {
+        Ok(self.read_bytes(1)?[0] != 0)
     }
 
     // Writes a UTF String into the buffer, with its length as a prefix
     pub fn write_utf(&mut self, utf_string: &str) -> Result<(), Error> {
         if let Ok(str_size) = u16::try_from(utf_string.len()) {
-            self.write_unsigned_short(str_size);
-            self.write_bytes(utf_string.as_bytes());
+            self.write_unsigned_short(str_size)?;
+            self.write_bytes(utf_string.as_bytes())
         } else {
-            return Err("RangeError: UTF String length must fit into a short".into());
+            Err("RangeError: UTF String length must fit into a short".into())
         }
-        Ok(())
-    }
-
-    pub fn get(&self, item: usize) -> Option<u8> {
-        self.bytes.get(item).copied()
-    }
-
-    pub fn get_range(&self, item: Range<usize>) -> Option<&[u8]> {
-        self.bytes.get(item)
     }
 
     pub fn set(&mut self, item: usize, value: u8) {
-        if self.bytes.len() < (item + 1) {
-            self.bytes.resize(item + 1, 0)
+        if self.len() < (item + 1) {
+            self.resize(item + 1, 0)
         }
 
-        *self.bytes.get_mut(item).unwrap() = value;
+        *self.get_mut(item).unwrap() = value;
     }
 
     pub fn delete(&mut self, item: usize) {
-        if let Some(i) = self.bytes.get_mut(item) {
+        if let Some(i) = self.get_mut(item) {
             *i = 0;
         }
     }
 
-    pub fn bytes(&self) -> &Vec<u8> {
-        &self.bytes
-    }
-
+    #[inline]
     pub fn position(&self) -> usize {
-        self.position
+        self.position.get()
     }
 
-    pub fn set_position(&mut self, pos: usize) {
-        self.position = pos;
+    #[inline]
+    pub fn set_position(&self, pos: usize) {
+        self.position.set(pos);
     }
 
-    pub fn add_position(&mut self, amnt: usize) {
-        self.position += amnt;
+    #[inline]
+    pub fn add_position(&self, amnt: usize) {
+        self.position.set(self.position.get() + amnt);
     }
 
+    #[inline]
     pub fn endian(&self) -> &Endian {
         &self.endian
     }
 
+    #[inline]
     pub fn set_endian(&mut self, new_endian: Endian) {
         self.endian = new_endian;
     }
 
-    pub fn len(&self) -> usize {
-        self.bytes.len()
+    pub fn bytes_available(&self) -> usize {
+        let pos = self.position.get();
+        if pos > self.len() {
+            0
+        } else {
+            self.len() - pos
+        }
+    }
+}
+
+impl std::ops::Deref for ByteArrayStorage {
+    type Target = Vec<u8>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.bytes
+    }
+}
+
+impl std::ops::DerefMut for ByteArrayStorage {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.bytes
     }
 }
```

Comment:
I don't think we should add `Deref` and `DerefMut` impls:
 * In general, Deref should be avoided because and only be used for pointer-like types. (see the [`Deref` docs](https://doc.rust-lang.org/std/ops/trait.Deref.html)) ByteArrayStorage isn't a smart pointer, but instead a 'wrapped' type, wrapping the behavior of Vec.
 * `Deref` itself might be okay, because you could view the storage as a pointer to a slice, but it should be of `type Target = [u8]`, because `&Vec<u8>` is almost always a mistake (see [this clippy lint](https://rust-lang.github.io/rust-clippy/master/#useless_vec)). I'd prefer an explicit `bytes()` or `as_slice() -> &[u8]` method.
 * `DerefMut` is definitely bad because there's no guarantees that invariants are upheld; outside code can fiddle with the vector however they want (For example, clearing it without updating the position).

In general, prefer adding dedicated methods on `ByteArrayStorage`: `clear`, `resize`, etc. It may be more typing, but it's better to be explicit.